### DISCUSSION TITLE
Voodoo 4 and 5 emulation

### DISCUSSION THREAD

**Discussion by Torinde** [PCBox/PCBox#89] [Ideas]
At https://github.com/86Box/86Box/issues/2710 there are:
- documents: [voodoo45_spec.pdf](https://github.com/86Box/86Box/files/9740186/voodoo45_spec.pdf), [voodoo45_spec.doc](https://vgamuseum.info/images/doc/3dfx/voodoo45_spec.doc), [voodoo45_databook.pdf](https://vgamuseum.info/images/doc/3dfx/voodoo45_databook.pdf)
- Voodoo5 [ROM v5500.zip](https://github.com/86Box/86Box/files/14893805/ROM.v5500.zip); Voodoo5 6000 has a [modern incarnation](https://www.modlabs.net/forum/topic/60674/), so surely its ROM was also dumped.

Per [Wikipedia](https://en.wikipedia.org/wiki/3dfx_Interactive#Products) the performance is:

- the already supported 3500SE: 400MTexels/s, 200MPixels/s
- Voodoo4 4500: 332MTexels/s (**less than 3500SE**), 332MPixels/s
- Voodoo5 5500: 664MTexels/s, 664MPixels/s
- Voodoo5 6000 166MHz: 1328MTexels/s, 1328MPixels/s
- Voodoo5 [6000 183MHz](https://en.wikipedia.org/wiki/Voodoo_5#Voodoo_5_6000): 1470MTexels/s, 1470MPixels/s

Per [PCBox docs](https://docs.pcbox-emu.xyz/en/latest/settings/display.html#voodoo-graphics) Voodoo emulation is already using multiple threads on the host. Modern top desktop CPUs have up to 32 threads (without going into Threadripper and server models). Voodoo5 utilizing multiple chips means the emulation should be parallelizable to multiple threads.

Per benchmarks shown [here](https://github.com/86Box/86Box/discussions/2909#discussioncomment-4498853) Games will benefit with CPU emulation speed matching Pentium II 450, so host requirements shouldn't be outrageous.

**Comment by bachimusprime** [comment]
A Ryzen 7 7700X is already just BARELY enough to run a Pentium 2 450 at full speed.

**Reply by Torinde** [reply]
The benchmark above uses P2/450 as the lowest for its results, but since it benefits _sufficiently_ from Voodoo5 5500, I assume even slower CPUs will also benefit.

And for:
> New (optional) dynamic recompiler that sacrifices timing accuracy for speed

If we take DOSbox as example, synthetic benchmarks [overtake P4/1.5GHz](https://github.com/dosbox-staging/dosbox-staging/pull/3426#issuecomment-1963060887):
> Of course this is just a synthetic benchmark, but I can clearly see that installing software (like Adobe Acrobat Reader, which took a lot of time to decompress) is crazy fast compared to what I remember from my old AMD K6-2 450 MHz.

**Reply by Coco19000s** [reply]
Voodoo 5 Emulation, even with a lower class Pentium mmx or Pentium 2, would still do very well as it was Very Powerful for the time.
